CASSANDRA-20752 Add  SSTableFlushObserver#onSSTableWriterSwitched

pranavshenoy opened the pull request:
add SSTableFlushObserver#onSSTableWriterSwitched to flush SAI segment builder for written shards without waiting for entire transaction to complete

This reduces memory usage during sharded compaction.


The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/segment/SegmentBuilder.java:
```diff
@@ -47,7 +47,7 @@ public abstract class SegmentBuilder
 
     /** The number of column indexes being built globally. (Starts at one to avoid divide by zero.) */
-    private static final AtomicInteger ACTIVE_BUILDER_COUNT = new AtomicInteger(0);
+    private static final AtomicInteger ACTIVE_BUILDER_COUNT = new AtomicInteger(1);
```
In the DS fork, we actually do set this to 1, but here in OSS, we start at zero and ensure safety by making sure the access and increment/decrement operations happen in the correct order below. I would revert this and remove the misleading field comment. However, that does mean you'll probably get a failure in `CompactionTest#testSegmentBuilderFlushWithShardedCompaction()`, which assumes we initialize to 1 (and this is correct in the DS fork, of course). That should be easy enough to fix in the test itself...

pranavshenoy replied:
done

maedhroz commented on test/unit/org/apache/cassandra/index/sai/functional/CompactionTest.java:
```diff
@@ -352,4 +359,53 @@ public void testConcurrentIndexDropWithCompaction() throws Throwable
+            while (!future.isDone())
+            {
+                // ACTIVE_BUILDER_COUNT starts from 1. There are 2 segments for 2 indexes
+                assertThat(SegmentBuilder.getActiveBuilderCount()).isGreaterThanOrEqualTo(1).isLessThanOrEqualTo(3);
```
If we revert the `ACTIVE_BUILDER_COUNT` thing above, this becomes...

```
// ACTIVE_BUILDER_COUNT starts from 0. There are 2 segments for 2 indexes
assertThat(SegmentBuilder.getActiveBuilderCount()).isGreaterThanOrEqualTo(0).isLessThanOrEqualTo(2);
```

(Suggested by Claude, but I think it's correct.)

pranavshenoy replied:
done. Thanks

maedhroz commented on test/unit/org/apache/cassandra/index/sai/functional/CompactionTest.java:
```diff
@@ -352,4 +359,53 @@ public void testConcurrentIndexDropWithCompaction() throws Throwable
+        finally
+        {
+            executor.shutdown();
+            executor.awaitTermination(30, TimeUnit.SECONDS);
```
nit: If this times out, we might want to fail the test?

maedhroz commented on test/unit/org/apache/cassandra/index/sai/functional/CompactionTest.java:
```diff
@@ -57,10 +62,12 @@
 import org.apache.cassandra.service.ActiveRepairService;
 import org.apache.cassandra.streaming.PreviewKind;
 import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.FBUtilities;
```
unused import

maedhroz commented on test/unit/org/apache/cassandra/io/sstable/SSTableFlushObserverTest.java:
```diff
@@ -32,10 +32,12 @@
 import org.apache.commons.lang3.tuple.ImmutableTriple;
 import org.apache.commons.lang3.tuple.Triple;
 import org.junit.Assert;
+import static org.junit.Assert.assertFalse;
```
unused import

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/SSTableIndexWriter.java:
```diff
@@ -100,6 +100,18 @@ public void addRow(PrimaryKey key, Row row, long sstableRowId) throws IOExceptio
+            return;
+
+        boolean emptySegment = currentBuilder == null || currentBuilder.isEmpty();
+        logger.debug("Flushing index {} with {}buffered data on sstable writer switched...", index.identifier().indexName, emptySegment ? "no " : "");
```
This should use the standard format...

```suggestion
        logger.debug(index.identifier().logMessage("Flushing index with {}buffered data on SSTable writer switched..."), emptySegment ? "no " : "");
```

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/v1/SSTableIndexWriter.java:
```diff
@@ -100,6 +100,18 @@ public void addRow(PrimaryKey key, Row row, long sstableRowId) throws IOExceptio
+        boolean emptySegment = currentBuilder == null || currentBuilder.isEmpty();
+        logger.debug("Flushing index {} with {}buffered data on sstable writer switched...", index.identifier().indexName, emptySegment ? "no " : "");
+        if (!emptySegment)
+            flushSegment();
```
Note that we don't have to aggressively create a new builder here because `addTerm()`, `complete()`, and `abort()` will all handle the `currentBuilder == null` case appropriately.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/PerColumnIndexWriter.java:
```diff
@@ -41,6 +41,12 @@ public interface PerColumnIndexWriter
     void complete(Stopwatch stopwatch) throws IOException;
 
+    /**
+     * Called when current sstable writer is switched during sharded compaction to free any in-memory resources associated
```
```suggestion
     * Called when current SSTable writer is switched during sharded compaction to free any in-memory resources associated
```

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/PerColumnIndexWriter.java:
```diff
@@ -41,6 +41,12 @@ public interface PerColumnIndexWriter
 
+    /**
+     * Called when current sstable writer is switched during sharded compaction to free any in-memory resources associated
+     * with the sstable for current index without waiting for full transaction to complete
```
```suggestion
     * with the SSTable for current index without waiting for full transaction to complete
```

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/StorageAttachedIndexWriter.java:
```diff
@@ -164,6 +164,27 @@ public void staticRow(Row staticRow)
+            logger.error(indexDescriptor.logMessage("Failed to flush segment on sstable writer switched"), t);
+            abort(t, true);
+            // fail compaction task or index build task if SAI failed
+            throw Throwables.unchecked(t);
```
This is unnecessary. When `abort()` is called with `fromIndex == true`, as we do here, the exception will be thrown at the end of `abort()`.

pranavshenoy replied:
done

maedhroz reviewed (changes requested):
Dropped some comments throughout the PR. We should probably also make sure we don't refer to CNDB-14725 in the commit message ;)

jasonstack reviewed (approved):
LGTM

maedhroz commented on test/unit/org/apache/cassandra/index/sai/functional/CompactionTest.java:
```diff
@@ -352,4 +358,54 @@ public void testConcurrentIndexDropWithCompaction() throws Throwable
+        {
+            executor.shutdown();
+            executor.awaitTermination(30, TimeUnit.SECONDS);
+            assertThat(executor.isTerminated()).isTrue();
```
nit: I guess what I meant was that you can just look at the return value from `awaitTermination()`. (Has the side effect of making the static analyzers happy...)

pranavshenoy replied:
done

maedhroz reviewed (approved):
Latest changes look good

pranavshenoy committed:
add SSTableFlushObserver#onSSTableWriterSwitched to flush SAI segment builder for written shards without waiting for entire transaction to complete. This reduces memory usage during sharded compaction.

    Flush segment builder when sstable writer is switched to free memory
    without waiting full compaction to complete

Patch by Pranav Shenoy and reviewed by Caleb Rackliffe and Zhao Yang

maedhroz commented:
Committed as https://github.com/apache/cassandra/commit/11278547e1c8b532969a81675750b2e542b0e379
